elastic/kibana
# [Execution context] Add nested context support

## Summary

Part of https://github.com/elastic/kibana/issues/102626

Adds support for declaring parent-child relationships between execution contexts. With this change, Kibana might log and pass to Elasticsearch information about the origin of every spawned execution context.
*kibana logs*
```json
{
   "parent":{
      "type":"application",
      "name":"dashboard",
      "id":"722b74f0-b882-11e8-a6d9-e546fe2bba5f",
      "description":"[eCommerce] Revenue Dashboard",
      "url":"/view/722b74f0-b882-11e8-a6d9-e546fe2bba5f"
   },
   "type":"visualization",
   "name":"metrics",
   "id":"45e07720-b890-11e8-a6d9-e546fe2bba5f",
   "description":"[eCommerce] Promotion Tracking",
   "url":"/app/visualize#/edit/45e07720-b890-11e8-a6d9-e546fe2bba5f"
}
```
*x-opaque-id* header
```
75221494-b6fb-40e2-a392-fdab0f500eb8;kibana:application:dashboard:722b74f0-b882-11e8-a6d9-e546fe2bba5f;visualization:metrics:45e07720-b890-11e8-a6d9-e546fe2bba5f
```

#### Client-side API
Due to lack of native JS API, we have to pass a parent context during instantiation of a child context:
```js
// in the parent app
const parentContext = {
  type: 'application',
  name: 'dashboard',
  ...
});

createChild(parentContext);

// in a child app
const context = executionContext.create({
  parent: parentContext,
  type: 'vis',
  ...
});
```

Note that `EmbeddableInput` is expected to be compatible with `SerializableState` interface, so I had to remove `IExecutionContextContainer` usage in favor of `KibanaExecutionContext` which is POJO.  With this change, using `execution_context` service on the client-side looks like overhead, so I removed it completely.

#### Server-side API
On the server-side, plugins may register `execution context` by wrapping any function in `withContext`. Provided context object is carried over all async operations spawned by the passed function. Sequentual calls `withContext` will create a stack of context.
```js
// in the parent app
executionContext.withContext({
  type: 'task',
  name: 'my-task',
  ...
}, async () => {
  return await fetchData();
});
```

### How to test
```yml
logging:
  appenders:
    myconsole:
      type: console
      layout:
        type: pattern
        pattern: '%date|%meta'
  loggers:
    - name: elasticsearch.query
      level: all
      appenders: [myconsole]
    - name: elasticsearch.query.taskManager
      level: off
    - name: execution_context
      level: debug
      appenders: [console]
```
- run Kibana
- navigate to Kibana Browser App
- Install sample data. I tested on `Sample eCommerce orders`.
- Navigate to dashboards.
- You should be able to see the output like this:
```
[2021-08-03T15:30:58.891+03:00][DEBUG][execution_context] set the execution context: {"parent":{"type":"application","name":"dashboard","id":"722b74f0-b882-11e8-a6d9-e546fe2bba5f","description":"[eCommerce] Revenue Dashboard","url":"/view/722b74f0-b882-11e8-a6d9-e546fe2bba5f"},"type":"visualization","name":"metrics","id":"45e07720-b890-11e8-a6d9-e546fe2bba5f","description":"[eCommerce] Promotion Tracking","url":"/app/visualize#/edit/45e07720-b890-11e8-a6d9-e546fe2bba5f"}

2021-08-03T15:30:58.899+03:00|{"http":{"request":{"id":"75221494-b6fb-40e2-a392-fdab0f500eb8;kibana:application:dashboard:722b74f0-b882-11e8-a6d9-e546fe2bba5f;visualization:metrics:45e07720-b890-11e8-a6d9-e546fe2bba5f"}}}
```

### Checklist

Delete any items that are not applicable to this PR.

### For maintainers
